# PART ONE #

# Question 1

### Total Scores for Students

Dataset is represented as matrix A

$ A = \begin{bmatrix}80&70&90\cr60 & 85 & 75\cr95 & 88 & 92\cr70 & 60 & 65\end{bmatrix}$

In the matrix A, the rows represent the student and the columns represent the subject.

- Student 1: \( 80 + 70 + 90 = 240 \)
- Student 2: \( 60 + 85 + 75 = 220 \)
- Student 3: \( 95 + 88 + 92 = 275 \)
- Student 4: \( 70 + 60 + 65 = 195 \)


In [ ]:
import numpy as np

A = np.array([[80, 70, 90], [60, 85, 75], [95, 88, 92], [70, 60, 65]])

print("Original Scores Matrix:\n ")
print(A)

total_scores = np.sum(A, axis=1)
print("\nTotal Scores per Student: ")
for i, total in enumerate(total_scores, 1):
    print(f"Student {i}: {total}")



Original Scores Matrix:
 
[[80 70 90]
 [60 85 75]
 [95 88 92]
 [70 60 65]]

Total Scores per Student: 
Student 1: 240
Student 2: 220
Student 3: 275
Student 4: 195


# Question 2

## Scalar Multiplication for Math

Weights: Math = 2, English = 1, Science = 1.

Weighted matrix As: 

$  As = \begin{bmatrix}80 * 2 & 70 * 1 & 90 * 1 \\60 * 2 & 85 * 1 & 75 * 1 \\95 * 2 & 88 * 1 & 92 * 1 \\70 * 2 & 60 * 1 & 65 * 1\end{bmatrix}$

$ As = \begin{bmatrix}160 & 70 & 90 \\120 & 85 & 75 \\190 & 88 & 92 \\140 & 60 & 65\end{bmatrix}$

- Weighted total scores:
  - Student 1: \( 160 + 70 + 90 = 320 \)
  - Student 2: \( 120 + 85 + 75 = 280 \)
  - Student 3: \( 190 + 88 + 92 = 370 \)
  - Student 4: \( 140 + 60 + 65 = 265 \)

In [ ]:
scaled_math = np.array([2, 1, 1])  
weighted_math = A * scaled_math
weighted_totals_math = np.sum(weighted_math, axis=1)
print("Weighted Scores Matrix (Maths twice important):")
print(weighted_math)
print("\nWeighted Total Scores:")
for i, total in enumerate(weighted_totals_math, 1):
    print(f"Student {i}: {total}")

Weighted Scores Matrix (Maths twice important):
[[160  70  90]
 [120  85  75]
 [190  88  92]
 [140  60  65]]

Weighted Total Scores:
Student 1: 320
Student 2: 280
Student 3: 370
Student 4: 265


# Question 3

## Subject Average

Subject Average = Total of Subject / No of Student

- Math: \( (80 + 60 + 95 + 70) / 4 = 305 / 4 = 76.25 \)
- English: \( (70 + 85 + 88 + 60) / 4 = 303 / 4 = 75.75 \)
- Science: \( (90 + 75 + 92 + 65) / 4 = 322 / 4 = 80.50 \)

In [ ]:
subjects = ['Math', 'English', 'Science']
subject_avg = np.mean(A, axis=0)
print("Average Score per Subject:")
for subject, avg in zip(subjects, subject_avg):
    print(f"{subject}: {avg:.2f}")


Average Score per Subject:
Math: 76.25
English: 75.75
Science: 80.50


# Question 4

## Final Grades

Final Grades: multiply through with weight vector [0.5, 0.3, 0.2]

- Student 1: \( 80 x 0.5 + 70 x 0.3 + 90 x 0.2 = 40 + 21 + 18 = 79 \)
- Student 2: \( 60 x 0.5 + 85 x 0.3 + 75 x 0.2 = 30 + 25.5 + 15 = 70.5 \)
- Student 3: \( 95 x 0.5 + 88 x 0.3 + 92 x 0.2 = 47.5 + 26.4 + 18.4 = 92.3 \)
- Student 4: \( 70 x 0.5 + 60 x 0.3 + 65 x 0.2 = 35 + 18 + 13 = 66 \)

In [ ]:
weights = np.array([0.5, 0.3, 0.2])
final_grades = np.matmul(A, weights)
print("Final Grades (weights: Math=0.5, English=0.3, Science=0.2):")
for i, grade in enumerate(final_grades, 1):
    print(f"Student {i}: {grade:.2f}")


Final Grades (weights: Math=0.5, English=0.3, Science=0.2):
Student 1: 79.00
Student 2: 70.50
Student 3: 92.30
Student 4: 66.00


# Question 5

## Comparing Students

Compare Student 1 and Student 2

- Student 1 vector: \( [80, 70, 90] \)
- Student 2 vector: \( [60, 85, 75] \)
- Overall Comparison = \([ 80+70+90 = 240 ] - [ 60+85+75 = 220] = 20 ) - Student 1 performed better

Subject Comparison
  - Math: \( 80-60 = 20 \) - Student 1 performed better
  - English: \( 70-85 = -15 \) - Student 2 performed better
  - Science: \( 90-75 = 15 \) - Student 1 performed better

In [ ]:
student1 = A[0]
student2 = A[1]
stu1 = np.sum(A[0])
stu2 = np.sum(A[1])
comp = stu1 - stu2
difference = student1 - student2

print("Overall Comparison: ")
if comp > 0:
    print("     Student 1 is better")
elif comp < 0:
    print("     Student 2 is better")
else:
    print("     Tie")

print("\nSubject Comparison:")
for subject, diff in zip(subjects, difference):
    print(f"{subject}: {diff}")
    if diff > 0:
        print("  Student 1 better")
    elif diff < 0:
        print("   Student 2 better")
    else:
        print("  Tie")
print()


Overall Comparison: 
     Student 1 is better

Subject Comparison:
Math: 20
  Student 1 better
English: -15
   Student 2 better
Science: 15
  Student 1 better



#

#

#

#

#

# PART TWO


# Question 1

### Total Scores for Students

Dataset is represented as matrix A

$ B = \begin{bmatrix}80 & 70 & 90 & 95\cr60 & 85 & 75 & 95\cr95 & 88 & 92 & 95\cr70 & 60 & 65 & 95\cr65 & 75 & 80 & 85\end{bmatrix}$

In the matrix A, the rows represent the student and the columns represent the subject.

- Student 1: \( 80 + 70 + 90 + 95 = 335 \)
- Student 2: \( 60 + 85 + 75 + 95 = 315 \)
- Student 3: \( 95 + 88 + 92 + 95 = 370 \)
- Student 4: \( 70 + 60 + 65 + 95 = 290 \)
- Student 5: \( 65 + 75 + 80 + 85 = 305 \)


In [ ]:
import numpy as np

B = np.array([[80, 70, 90, 95], [60, 85, 75, 95], [95, 88, 92, 95], [70, 60, 65, 95], [65, 75, 80, 85]])

print("Original Scores Matrix:\n ")
print(B)

total_scores = np.sum(B, axis=1)
print("\nTotal Scores per Student: ")
for i, total in enumerate(total_scores, 1):
    print(f"Student {i}: {total}")



Original Scores Matrix:
 
[[80 70 90 95]
 [60 85 75 95]
 [95 88 92 95]
 [70 60 65 95]
 [65 75 80 85]]

Total Scores per Student: 
Student 1: 335
Student 2: 315
Student 3: 370
Student 4: 290
Student 5: 305


# Question 2

## Scalar Multiplication for Math

Weights: Math = 1, English = 1, Science = 3, History = 1.

Weighted matrix Bs: 

$ Bs = \begin{bmatrix}80*1 & 70*1 & 90*3 & 95*1\cr60*1 & 85*1 & 75*3 & 95*1\cr95*1 & 88*1 & 92*3 & 95*1\cr70*1 & 60*1 & 65*3 & 95*1\cr65*1 & 75*1 & 80*3 & 85*1\end{bmatrix}$

$ As = \begin{bmatrix}80 & 70 & 270 & 95\\60 & 85 & 225 & 95\\95 & 88 & 276 & 95\\70 & 60 & 195 & 95\\65 & 75 & 240 & 85\end{bmatrix}$

- Weighted total scores:
  - Student 1: \( 80 + 70 + 270 + 95 = 515 \)
  - Student 2: \( 60 + 85 + 225 + 95 = 465 \)
  - Student 3: \( 95 + 88 + 276 + 95 = 554 \)
  - Student 4: \( 70 + 60 + 195 + 95 = 420 \)
  - Student 5: \( 65 + 75 + 240 + 85 = 465 \)

In [ ]:
scaled_maths = np.array([1, 1, 3, 1])  
weighted_maths = B * scaled_maths
weighted_total_maths = np.sum(weighted_maths, axis=1)
print("Weighted Scores Matrix (Science Thrice Important):")
print(weighted_maths)
print("\nWeighted Total Scores:")
for i, total in enumerate(weighted_total_maths, 1):
    print(f"Student {i}: {total}")

Weighted Scores Matrix (Science Thrice Important):
[[ 80  70 270  95]
 [ 60  85 225  95]
 [ 95  88 276  95]
 [ 70  60 195  95]
 [ 65  75 240  85]]

Weighted Total Scores:
Student 1: 515
Student 2: 465
Student 3: 554
Student 4: 420
Student 5: 465


# Question 3

## Final Grades

Final Grades: multiply through with weight vector [0.4, 0.3, 0.2, 0.1]

- Student 1: \( 80 x 0.4 + 70 x 0.3 + 90 x 0.2 + 95 x 0.1 = 32 + 21 + 18 + 9.5 = 80.5 \)
- Student 2: \( 60 x 0.4 + 85 x 0.3 + 75 x 0.2 + 95 x 0.1 = 24 + 25.5 + 15 + 9.5 = 74 \)
- Student 3: \( 95 x 0.4 + 88 x 0.3 + 92 x 0.2 + 95 x 0.1 = 38 + 26.4 + 18.4 + 9.5 = 92.3 \)
- Student 4: \( 70 x 0.4 + 60 x 0.3 + 65 x 0.2 + 95 x 0.1 = 28 + 18 + 13 + 9.5 = 68.5 \)
- Student 5: \( 65 x 0.4 + 75 x 0.3 + 80 x 0.2 + 85 x 0.1 = 26 + 22.5 + 16 + 8.5 = 73)

In [ ]:
weight = np.array([0.4, 0.3, 0.2, 0.1])
final_grade = np.matmul(B, weight)
print("Final Grades (weights: Math=0.4, English=0.3, Science=0.2, History=0.1):")
for i, grades in enumerate(final_grade, 1):
    print(f"Student {i}: {grades:.2f}")


Final Grades (weights: Math=0.4, English=0.3, Science=0.2, History=0.1):
Student 1: 80.50
Student 2: 74.00
Student 3: 92.30
Student 4: 68.50
Student 5: 73.00


# Question 4

## Comparing Students

Compare Student 3 and Student 4

- Student 3: \( 95 + 88 + 92 + 95 = 370 \)
- Student 4: \( 70 + 60 + 65 + 95 = 290 \)
- Overall Comparison = \( 95 + 88 + 92 + 95 = 370 \) - \( 70 + 60 + 65 + 95 = 290 \) = 80 - Student 3 performed better

Subject Comparison
  - Math: \( 95 - 70 = 25 \) - Student 3 performed better
  - English: \( 88 - 60 = 28 \) - Student 3 performed better
  - Science: \( 92 - 65 = 27 \) - Student 3 performed better
  - History: \( 95 - 95 = 0 \) - TIE

In [ ]:
student3 = B[2]
student4 = B[3]
stu3 = np.sum(B[2])
stu4 = np.sum(B[4])
comp = stu3 - stu4
difference = student3 - student4

print("Overall Comparison: ")
if comp > 0:
    print("     Student 3 is better")
elif comp < 0:
    print("     Student 4 is better")
else:
    print("     Tie")
    
subj = ['Math', 'English', 'Science', 'History']
print("\nSubject Comparison:")
for subject, diff in zip(subj, difference):
    print(f"{subject}: {diff}")
    if diff > 0:
        print("  Student 3 better")
    elif diff < 0:
        print("  Student 4 better")
    else:
        print("  Tie")
print()

Overall Comparison: 
     Student 3 is better

Subject Comparison:
Math: 25
  Student 3 better
English: 28
  Student 3 better
Science: 27
  Student 3 better
History: 0
  Tie

